src/main/java/com/ebay/myriad/scheduler/NMHeartBeatHandler.java

```diff
@@ -0,0 +1,154 @@
+        yarnNodeCapacityMgr.removeCapacity(rmNode);
+      }
+      break;
+
+      case EXPIRE: {
+        nodeStore.remove(event.getNodeId().getHost());
+      }
+      break;
+
+      case STATUS_UPDATE: {
+        handleStatusUpdate(event, context);
+      }
+      break;
+
+      default: {
+        LOGGER.debug("Unhandled event: {}", event.getClass().getName());
+      }
+      break;
+    }
+  }
+
+  private void handleStatusUpdate(RMNodeEvent event, RMContext context) {
+    if (!(event instanceof RMNodeStatusEvent)) {
+      LOGGER.error("{} not an instance of {}", event.getClass().getName(),
+          RMNodeStatusEvent.class.getName());
+      return;
+    }
+
+    RMNodeStatusEvent statusEvent = (RMNodeStatusEvent) event;
+    RMNode rmNode = context.getRMNodes().get(event.getNodeId());
+    SchedulerNode schedulerNode = yarnScheduler.getSchedulerNode(rmNode.getNodeID());
+    String hostName = rmNode.getNodeID().getHost();
+
+    int rmSchNodeNumContainers = schedulerNode.getNumContainers();
+    List<ContainerStatus> nmContainers = statusEvent.getContainers();
+
+    if (nmContainers.size() != rmSchNodeNumContainers) {
+      LOGGER.warn("Node: {}, Num Containers known by RM scheduler: {}, "
+          + "Num containers NM is reporting: {}",
+          hostName, rmSchNodeNumContainers, nmContainers.size());
+    }
+
+    // Send task update to Mesos
+    for (ContainerStatus status : nmContainers) {
+      ContainerId containerId = status.getContainerId();
+      Protos.SlaveID slaveId = nodeStore.getNode(hostName).getSlaveId();
+      if (status.getState() == ContainerState.COMPLETE) {
+        LOGGER.debug("Task complete: {}", containerId);
+        yarnNodeCapacityMgr.removeCapacity(rmNode, containerId);
+        requestExecutorToSendTaskStatusUpdate(
+            slaveId,
+            containerId, Protos.TaskState.TASK_FINISHED);
+      } else { // state == NEW | RUNNING
+        requestExecutorToSendTaskStatusUpdate(
+            slaveId,
+            containerId, Protos.TaskState.TASK_RUNNING);
+      }
+    }
+  }
+
+  /**
+   * sends a request to executor on the given slave to send back a status update
+   * for the mesos task launched for this container.
+   *
+   * @param slaveId
+   * @param containerId
+   * @param taskState
+   */
+  private void requestExecutorToSendTaskStatusUpdate(Protos.SlaveID slaveId,
+      ContainerId containerId,
+      Protos.TaskState taskState) {
+    ContainerTaskStatusRequest containerTaskStatusRequest = new ContainerTaskStatusRequest();
+    containerTaskStatusRequest.setMesosTaskId(
+        ContainerTaskStatusRequest.YARN_CONTAINER_TASK_ID_PREFIX + containerId.toString());
+    containerTaskStatusRequest.setState(taskState.name());
+    myriadDriver.getDriver().sendFrameworkMessage(
+        getExecutorId(slaveId),
+        slaveId,
+        new Gson().toJson(containerTaskStatusRequest).getBytes(Charset.defaultCharset()));
+  }
```

Comment:
`sendFrameworkMessage` is [unreliable](http://mesos.apache.org/api/latest/java/org/apache/mesos/SchedulerDriver.html#sendFrameworkMessage%28ExecutorID,%20SlaveID,%20byte[]%29), is there any redundancy here?  Otherwise if the message is dropped we'll never drop the capacity.  Slow resource leak?